fix: remove outdated ipfs gateways

jpcloureiro opened the pull request:
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/coding_guidelines/CODING_GUIDELINES.md)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add the appropiate QA label when dev review is completed
    - `needs-qa`: PR requires manual QA.
    - `No QA/E2E only`: PR does not require any manual QA effort. Prior to merging, ensure that you have successful end-to-end test runs in Bitrise. 
    - `Spot check on release build`: PR does not require feature QA but needs non-automated verification. In the description section, provide test scenarios. Add screenshots, and or recordings of what was tested.
5. Add `QA Passed` label when QA has signed off (Only required if the PR was labeled with `needs-qa`)
6. Add your team's label, i.e. label starting with `team-` (or `external-contributor` label if your not a MetaMask employee)

**Description**

Remove outdated ipfs gateways from selectable list under advanced settings

migrate users from outdated ipfs gateways URLs to the default one <TBA>


**Issue**

fixes https://github.com/MetaMask/mobile-planning/issues/1152

**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [ ] Any added code is fully documented

jpcloureiro committed:
migration: replace outdated ipfs gateway with default one

jpcloureiro committed:
migration: update first unit test

jpcloureiro committed:
ipfs gateways list: remove outdated URLs

jpcloureiro committed:
ipfs gateways: set cloudflare as default

jpcloureiro commented on app/store/migrations.js:
```diff
@@ -433,6 +433,24 @@ export const migrations = {
+    ];
+
+    const isUsingOutdatedGateway = outdatedIpfsGateways.includes(
+      state.engine.backgroundState.PreferencesController.ipfsGateway,
```
use optional chaining operator to avoid app crash if Pref Controller is not defined

Cal-L replied:
Sounds like you'll be updating this to use optional chaining?

jpcloureiro replied:
done https://github.com/MetaMask/metamask-mobile/pull/6962/commits/b1f2f230bcc52a65900f7404c382781b977933fc

jpcloureiro committed:
Merge branch 'main' into cirt/170-ipfs-gateway

Cal-L commented on app/store/migrations.test.js:
```diff
@@ -45,4 +49,46 @@ describe('Redux Persist Migrations', () => {
+      });
+
+      // State with outdated ipfs gateway
+      const currentState = stateWithIpfsGateway('https://hardbin.com/ipfs/');
```
Can we consolidate the IPFS URLs in a single place

jpcloureiro replied:
https://github.com/MetaMask/metamask-mobile/pull/6962#discussion_r1296055722

Cal-L commented on app/store/migrations.js:
```diff
@@ -433,6 +433,24 @@ export const migrations = {
   },
+  20: (state) => {
+    const outdatedIpfsGateways = [
+      'https://hardbin.com/ipfs/',
```
Can we consolidate the IPFS URLs in a single place

jpcloureiro replied:
Would this URL list be used elsewhere but for this migration & its unit test?

Cal-L replied:
Ah, I meant refer to the urls in `ipfs-gateways.json`. But then again those are already in its own json file so as long as those values aren't used elsewhere, it can be considered to be consolidated.

Cal-L commented:
@jpcloureiro Could you provide videos demonstrating the last two scenarios in the acceptance criteria?

jpcloureiro committed:
migration: return same state if any obj prop is undefined

Make sure we return the same state if ...PreferencesController or
...PreferencesController.ipfsGateway happen to be undefined for some
reason

jpcloureiro commented:
> @jpcloureiro Could you provide videos demonstrating the last two scenarios in the acceptance criteria?

`Scenario: Upgrading from old version with offline gateway selected should migrate over to use https://cloudflare-ipfs.com/ipfs/ by default`

https://recordit.co/nJruefTBl1

`Scenario: Upgrading from old version without online gateway selected should persist selected IPFS gateway in upgraded app`

https://recordit.co/OnMq5d33VF

They're also represented by [the unit tests](https://github.com/MetaMask/metamask-mobile/pull/6962/files#diff-1a29e4f64b7958d5ce8b2947fb286b8aaf6d9078c2cdd8c53008f870ee9affee)

Cal-L committed:
Merge branch 'main' into cirt/170-ipfs-gateway

jpcloureiro committed:
Merge branch 'main' into cirt/170-ipfs-gateway

tommasini reviewed (dismissed):
LGTM!

jpcloureiro committed:
Merge branch 'main' into cirt/170-ipfs-gateway

jpcloureiro committed:
Merge branch 'main' into cirt/170-ipfs-gateway

tommasini reviewed (approved):
LGTM!

cortisiko committed:
Merge branch 'main' into cirt/170-ipfs-gateway

cortisiko reviewed (approved):
✅ ✅
